### PR TITLE
Bug 2116686, Bug OCPBUGS-1493: Initial implementation of validation webhooks

### PR DESCRIPTION
**Pull request by Fedosin**
This PR starts the implementation of the validation webhooks for CAPI objects.
In particular, it validates that Cluster object infrastructure reference has correct value.

### CODE REVIEW

**Review by elmiko — COMMENTED** [review]
i couldn't tell if the test failures are related to this pr, it didn't look like it, but this generally makes sense to me.
/lgtm

File: pkg/webhook/cluster.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *ClusterWebhook) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	cluster, ok := obj.(*v1beta1.Cluster)
+	if !ok {
+		panic("expected to get an of object of type v1beta1.Cluster")
+	}
+	switch cluster.Spec.InfrastructureRef.Kind {
+	case "AWSCluster":
+	case "AzureCluster":
+	case "GCPCluster":
```
Comment (JoelSpeed) [inline pkg/webhook/cluster.go:34]: Nit, can inline these
```suggestion
	case "AWSCluster", "AzureCluster", "GCPCluster":
```

File: pkg/webhook/cluster.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	}
+	newCluster, ok := newObj.(*v1beta1.Cluster)
+	if !ok {
+		panic("expected to get an of object of type v1beta1.Cluster")
+	}
+
+	switch newCluster.Spec.InfrastructureRef.Kind {
+	case "AWSCluster":
+	case "AzureCluster":
+	case "GCPCluster":
```
Comment (JoelSpeed) [inline pkg/webhook/cluster.go:56]: Nit, inline please

File: pkg/webhook/coreprovider.go [hunk 3]
```diff
@@ -24,15 +24,37 @@ var _ webhook.CustomValidator = &CoreProviderWebhook{}
...
+		return fmt.Errorf("incorrect core provider name: %s", newCoreProvider.Name)
+	}
+
 	return nil
 }
 
 // ValidateDelete implements webhook.Validator so a webhook will be registered for the type
 func (r *CoreProviderWebhook) ValidateDelete(_ context.Context, obj runtime.Object) error {
-	return errors.New("deletion of infrastructure provider is not allowed")
+	return fmt.Errorf("deletion of infrastructure provider is not allowed")
```
Comment (JoelSpeed) [inline pkg/webhook/coreprovider.go:59]: You aren't using a format string so this should be `errors.New`

File: pkg/webhook/infraprovider.go [hunk 4]
```diff
@@ -24,15 +24,37 @@ var _ webhook.CustomValidator = &InfrastructureProviderWebhook{}
...
+		return fmt.Errorf("incorrect infra provider name: %s", newInfraProvider.Name)
+	}
+
 	return nil
 }
 
 // ValidateDelete implements webhook.Validator so a webhook will be registered for the type
 func (r *InfrastructureProviderWebhook) ValidateDelete(_ context.Context, obj runtime.Object) error {
-	return errors.New("deletion of infrastructure provider is not allowed")
+	return fmt.Errorf("deletion of infrastructure provider is not allowed")
```
Comment (JoelSpeed) [inline pkg/webhook/infraprovider.go:59]: You aren't using a format string so this should be `errors.New`

**Review by JoelSpeed — COMMENTED** [review]
I'd like to see some tests added for this before we merge. Please take a look at the CPMS webhook test for some inspiration on how we might do this.

It would also be good to set up some linting here, again, similar to CPMS

**Review by JoelSpeed — COMMENTED** [review]
/approve
/lgtm

**Comment by Fedosin** [comment]
/hold cancel

seems working

File: manifests/0000_30_cluster-api_10_webhooks.yaml [hunk 5]
```diff
@@ -49,3 +52,25 @@ webhooks:
         resources:
           - infrastructureproviders
     sideEffects: None
+  - admissionReviewVersions:
+      - v1
+      - v1beta1
```
Comment (JoelSpeed) [inline manifests/0000_30_cluster-api_10_webhooks.yaml:57]: Do we need both versions or should we just use `v1`?

Comment (Fedosin) [reply manifests/0000_30_cluster-api_10_webhooks.yaml:57]: In this cluster-api [example](https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/webhooks/cluster.go#L49) they recommend to use both versions.

**Comment by JoelSpeed** [comment]
/hold this is legitimately broken

The test suite downloads the latest version of the CLI instead of building from the source vendored into the project, this causes an issue because when new flags are added, the CLI sends new options to the library which then errors out. 

We need to make sure the CLI and library version are consistent.

**Comment by Fedosin** [comment]
@JoelSpeed I proposed a fix for CI: https://github.com/openshift/cluster-capi-operator/pull/80

**Pull request merged**